File: Tree/94.c
#include <stdio.h>
#include <stdlib.h>

// --- TreeNode structure ---
struct TreeNode {
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
};

struct TreeNode* newNode(int val) {
    struct TreeNode* node = (struct TreeNode*)malloc(sizeof(struct TreeNode));
    node->val = val;
    node->left = node->right = NULL;
    return node;
}

// --- Function declaration ---
void dfs(struct TreeNode* node, int* arr, int* idx){
    if(!node){
        return;
    }

    dfs(node->left, arr, idx);
    arr[*idx] = node->val;
    (*idx)++;
    dfs(node->right, arr, idx); 
}



int* inorderTraversal(struct TreeNode* root, int* returnSize){
   int* arr = malloc(sizeof(int)*256);
   *returnSize = 0;
   dfs(root, arr, returnSize);
   return arr;
}

int main() {
    // 建立測試樹：
    //        1
    //         \
    //          2
    //         /
    //        3
    struct TreeNode* root = newNode(1);
    root->right = newNode(2);
    root->right->left = newNode(3);

    int returnSize;
    int* result = inorderTraversal(root, &returnSize);

    printf("Inorder Traversal: ");
    for (int i = 0; i < returnSize; i++) {
        printf("%d ", result[i]);
    }
    printf("\n");

    // 若你在程式中用 malloc，記得釋放記憶體
    free(result);
    return 0;
}

File: Tree/94.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def inorderTraversal(self, root):
        result = []
        
        def dfs(node):
            if not node:
                return
            
            # 左
            dfs(node.left)
            
            # 自己
            result.append(node.val)
            
            # 右
            dfs(node.right)
        
        dfs(root)
        return result
    
def build_test_tree():
    """
          1
           \
            2
           /
          3

    inorder 應該是 [1,3,2]
    """
    root = TreeNode(1)
    root.right = TreeNode(2)
    root.right.left = TreeNode(3)
    return root


if __name__ == "__main__":
    root = build_test_tree()
    sol = Solution()
    print(sol.inorderTraversal(root))  # [1, 3, 2]